<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Joshua D. Welch and Velina Kozareva" />

<meta name="date" content="2023-11-08" />

<title>Comparing and contrasting heterogeneous single cell profiles using liger</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Comparing and contrasting heterogeneous
single cell profiles using liger</h1>
<h4 class="author">Joshua D. Welch and Velina Kozareva</h4>
<h4 class="date">2023-11-08</h4>



<div id="data-preprocessing" class="section level2">
<h2>Data preprocessing</h2>
<p>The algorithm takes a list of two or more digital gene expression
(DGE) matrices as input. Genes should be in rows and cells in columns.
Before running the factorization, we need to normalize the data to
account for different numbers of UMIs per cell, select variable genes,
and scale the data. Note that we do not center the data because
nonnegative matrix factorization accepts only positive values. The
selectGenes function performs variable gene selection on each of the
datasets separately, then takes the union. Note that corresponding genes
in each dataset need to have the same names (though the genes do not
need to be in the same order in each dataset). For cross-species
analysis, it may be convenient to convert all gene names to uppercase;
you can do this using the capitalize=T option of the selectGenes
function.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>dge1 <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;dge1.RDS&quot;</span>) <span class="co">#genes in rows, cells in columns, rownames and colnames included. Sparse matrix format is</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a>recommended.</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>dge2 <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;dge2.RDS&quot;</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">createLiger</span>(<span class="fu">list</span>(<span class="at">name1 =</span> dge1, <span class="at">name2 =</span> dge2)) <span class="co">#Can also pass in more than 2 datasets</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">normalize</span>(ligerex)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">selectGenes</span>(ligerex, <span class="at">var.thresh =</span> <span class="fl">0.1</span>)</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">scaleNotCenter</span>(ligerex)</span></code></pre></div>
<div id="loading-10x-data" class="section level3">
<h3>Loading 10X Data</h3>
<p><code>liger</code> also has functions for reading data generated by
10X’s <code>cellranger count</code> pipeline (including from 10X V3). We
can merge data by data type (most commonly Gene Expression) across
multiple samples and then use this as a single dataset in a new object
for integration.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># 10X data to be combined for sample 1 </span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>sample1_data1 <span class="ot">=</span> <span class="st">&quot;/path/to/10X/output/dir1&quot;</span></span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>sample1_data2 <span class="ot">=</span> <span class="st">&quot;/path/to/10X/output/dir2&quot;</span></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>sample1_dge <span class="ot">=</span> <span class="fu">read10X</span>(<span class="at">sample.dirs =</span> <span class="fu">list</span>(sample1_data1, sample1_data2), </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample.names =</span> <span class="fu">c</span>(<span class="st">&quot;s1_data1&quot;</span>, <span class="st">&quot;s1_data2&quot;</span>), <span class="at">min.umis =</span> <span class="dv">500</span>)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># 10X data for sample 2</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>sample2_data <span class="ot">=</span> <span class="st">&quot;/path/to/10X/output/dir3&quot;</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>sample2_dge <span class="ot">=</span> <span class="fu">read10X</span>(<span class="at">sample.dirs =</span> <span class="fu">list</span>(sample2_data), </span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample.names =</span> <span class="fu">c</span>(<span class="st">&quot;s2_data&quot;</span>), <span class="at">min.umis =</span> <span class="dv">500</span>)</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>liger10X <span class="ot">=</span> <span class="fu">createLiger</span>(<span class="fu">list</span>(<span class="at">sample1 =</span> sample1_dge, <span class="at">sample2 =</span> sample2_dge))</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># continue with other preprocessing steps</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>                      </span></code></pre></div>
</div>
</div>
<div id="performing-the-factorization" class="section level2">
<h2>Performing the factorization</h2>
<p>Next we perform the factorization using an alternating least squares
algorithm. After performing the factorization, we identify cells that
load on corresponding cell factors and quantile normalize their factor
loadings across datasets. The key parameters here are the number of
factors (k), the penalty parameter (lambda), and the clustering
resolution. In most cases, the default settings of lambda=5.0 and
resolution=1.0 provide reasonable results.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">optimizeALS</span>(ligerex, <span class="at">k =</span> <span class="dv">20</span>) </span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">quantileAlignSNF</span>(ligerex) <span class="co">#SNF clustering and quantile alignment</span></span></code></pre></div>
</div>
<div id="visualizing-the-results" class="section level2">
<h2>Visualizing the results</h2>
<p>We can visualize the results by using dimensionality reduction
techniques like t-SNE or UMAP (recommended for larger datasets).
Visualizations can be colored by dataset of origin, cluster assignment,
or any feature included in or added to the cell metadata
(<code>cell.data</code>).</p>
<p><code>plotWordClouds</code> and <code>plotGeneLoadings</code> are
useful ways to visualize the most highly loading genes (both shared and
dataset specific) for each factor, in conjunction with the factor
loadings across cells. These functions are very similar, but
<code>plotGeneLoadings</code> displays the top loading genes in ordered
scatter plots instead of word clouds.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">runTSNE</span>(ligerex)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># for larger datasets, may want to use UMAP instead</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">runUMAP</span>(ligerex)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">plotByDatasetAndCluster</span>(ligerex) <span class="co">#Can also pass in different set of cluster labels to plot</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">plotFeature</span>(ligerex, <span class="st">&quot;nUMI&quot;</span>)</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;word_clouds.pdf&quot;</span>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotWordClouds</span>(ligerex)</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;gene_loadings.pdf&quot;</span>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGeneLoadings</span>(ligerex)</span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
</div>
<div id="exploring-factors-and-clusters" class="section level2">
<h2>Exploring factors and clusters</h2>
<p>Another way to examine factor loadings across cells, and to help
visualize the alignment process is through <code>plotFactors</code>;
this can be helpful for seeing significant scale differences between the
datasets. We can also visualize the correspondence between clusters and
factors in the data with <code>plotClusterProportions</code> and
<code>plotClusterFactors</code>. These can be especially useful for
identifying clusters associated with certain factors and corresponding
marker genes.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotFactors</span>(ligerex)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotClusterProportions</span>(ligerex)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotClusterFactors</span>(ligerex, <span class="at">use.aligned =</span> T)</span></code></pre></div>
</div>
<div id="finding-and-visualizing-marker-genes" class="section level2">
<h2>Finding and visualizing marker genes</h2>
<p>We can use the factorization to more explicitly identify shared and
dataset-specific markers. The function <code>getFactorMarkers</code>
returns a list, where the first element contains dataset-specific
markers for dataset 1, the second element contains sharedmarkers, the
third element contains dataset-specific markers for dataset 2, and the
last 2 elements indicate the number of factors in which each marker is
found. This information allows the identification of ubiquitous
vs. cell-type-specific dataset differences.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>markers <span class="ot">=</span> <span class="fu">getFactorMarkers</span>(ligerex, <span class="at">num.genes =</span> <span class="dv">10</span>)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGene</span>(ligerex, <span class="at">gene =</span> <span class="st">&quot;Malat1&quot;</span>)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">plotGeneViolin</span>(ligerex, <span class="at">gene =</span> <span class="st">&quot;Malat1&quot;</span>)</span></code></pre></div>
</div>
<div id="comparing-different-cluster-assignments" class="section level2">
<h2>Comparing different cluster assignments</h2>
<p>We can compare and visualize <code>liger</code> cluster assignments
with existing clusterings (if cluster assignments are available for the
individual datasets).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># published cluster assignments for all cells in dataset 1 and 2</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>clusters_published1 <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;clusters1.RDS&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>clusters_published2 <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="st">&quot;clusters2.RDS&quot;</span>)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate adjusted Rand Index between liger cluster assignments and another assignment</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="fu">calcARI</span>(ligerex, <span class="fu">c</span>(clusters_published1, clusters_published2)) </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate purity between liger cluster assignments and another assignment for just dataset 1 </span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="fu">calcPurity</span>(ligerex, clusters_published1)</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co"># visualize joint cluster assignment as related to individual dataset cluster assignments</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="fu">makeRiverplot</span>(ligerex, clusters_published1, clusters_published2)</span></code></pre></div>
</div>
<div id="checking-dataset-alignment-and-individual-dataset-distortion" class="section level2">
<h2>Checking dataset alignment and individual dataset distortion</h2>
<p><code>liger</code> includes methods for estimating the level of
integration (alignment) between datasets and the level of distortion of
structure for each individual dataset after factorization and alignment
(compared to before). In general, datasets which have been factorized
and aligned in a meaningful way should show a high degree of integration
(high alignment metric), while maintaining a low degree of distortion
(high agreement metric).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">calcAlignment</span>(ligerex)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">calcAgreement</span>(ligerex)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># see if certain clusters are more integrated than others</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="fu">calcAlignmentPerCluster</span>(ligerex)</span></code></pre></div>
</div>
<div id="selecting-k-and-lambda" class="section level2">
<h2>Selecting k and lambda</h2>
<p>The <code>suggestK</code> and <code>suggestLambda</code> functions
can aid in selecting k and lambda. We want to find the smallest k for
which the increase in entropy metric begins to level off (an “elbow” in
the plot). Similarly, we want the smallest lambda for which the
alignment metric stabilizes.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suggestK</span>(ligerex) <span class="co"># plot entropy metric to find an elbow that can be used to select the number of factors</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">suggestLambda</span>(ligerex, k) <span class="co"># plot alignment metric to find an elbow that can be used to select the value of lambda</span></span></code></pre></div>
</div>
<div id="updating-the-factorization" class="section level2">
<h2>Updating the factorization</h2>
<p>If we want to add new data, change k or lambda, or re-analyze a
subset of the data, the functions below provide an efficient method of
updating. This is much faster than the naive approach of simply
re-running the optimizeALS algorithm.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">optimizeNewK</span>(ligerex, <span class="at">k =</span> <span class="dv">15</span>) <span class="co">#Can also decrease K</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="co">#Add new batches from the same condition/technology/species/protocol</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">optimizeNewData</span>(ligerex, <span class="at">newdata =</span> <span class="fu">list</span>(<span class="at">name1 =</span> dge1.new, <span class="at">name2 =</span> dge2.new),</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                          <span class="at">which.datasets =</span> <span class="fu">list</span>(name1, name2), <span class="at">add.to.existing =</span> T) </span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#Add completely new datasets. Specify which existing datasets are most similar.</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">optimizeNewData</span>(ligerex, <span class="at">newdata =</span> <span class="fu">list</span>(<span class="at">name3 =</span> dge3, <span class="at">name4 =</span> dge4),</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>                          <span class="at">which.datasets =</span> <span class="fu">list</span>(name1, name2), <span class="at">add.to.existing =</span> F) </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co">#cell.subset is a list of cells to retain from each dataset</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">optimizeSubset</span>(ligerex, cell.subset) </span></code></pre></div>
</div>
<div id="integration-between-liger-and-seurat" class="section level2">
<h2>Integration between liger and Seurat</h2>
<p>We can easily create <code>liger</code> objects from Seurat (V2 or
V3) objects (and vice versa), while keeping calculated features from the
original objects.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create liger object from two separate Seurat objects, keeping union of top 2000 highly variable </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co"># genes from each object</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">seuratToLiger</span>(<span class="fu">list</span>(seurat1, seurat2), <span class="at">names =</span> <span class="fu">c</span>(<span class="st">&#39;name1&#39;</span>, <span class="st">&#39;name2&#39;</span>), <span class="at">num.hvg.info =</span> <span class="dv">2000</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create liger object from single integrated Seurat V2 object, keeping CCA factorization, </span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co"># splitting datasets by Seurat meta.var column &quot;original&quot;</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">seuratToLiger</span>(seurat_obj, <span class="at">combined.seurat =</span> T, <span class="at">meta.var =</span> <span class="st">&#39;original&#39;</span>, <span class="at">cca.to.H =</span> T)</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create liger object from single integrated Seurat V3 object, splitting datasets by two</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true" tabindex="-1"></a><span class="co"># available assays in Seurat </span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true" tabindex="-1"></a>ligerex <span class="ot">=</span> <span class="fu">seuratToLiger</span>(seurat_obj, <span class="at">combined.seurat =</span> T, <span class="at">assays.use =</span> <span class="fu">c</span>(<span class="st">&#39;RNA&#39;</span>, <span class="st">&#39;ADT&#39;</span>))</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Create Seurat object from liger object, keeping liger highly variable genes</span></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true" tabindex="-1"></a>seurat_obj <span class="ot">=</span> <span class="fu">ligerToSeurat</span>(ligerex, <span class="at">use.liger.genes =</span> T)</span></code></pre></div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
