<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Joshua D. Welch" />

<meta name="date" content="2024-04-04" />

<title>Data Integration with LIGER</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>







<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Data Integration with LIGER</h1>
<h4 class="author">Joshua D. Welch</h4>
<h4 class="date">2024-04-04</h4>



<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>LIGER was initially introduced in <a href="https://doi.org/10.1016/j.cell.2019.05.006">Welch et al.Â 2019</a>
as a method for integrating single-cell RNA-seq data across multiple
technologies, species, and conditions. The method relies on integrative
nonnegative matrix factorization (iNMF) to identify shared and
dataset-specific factors.</p>
<p>LIGER can be used to compare and contrast experimental datasets in a
variety of contexts, for instance:</p>
<ul>
<li>Across experimental batches</li>
<li>Across individuals</li>
<li>Across sex</li>
<li>Across tissues</li>
<li>Across species (e.g., mouse and human)</li>
<li>Across modalities (e.g., scRNAseq and spatial transcriptomics data,
scMethylation, or scATAC-seq)</li>
</ul>
<p>Once multiple datasets are integrated, the package provides
functionality for further data exploration, analysis, and visualization.
Users can:</p>
<ul>
<li>Identify clusters</li>
<li>Find significant shared (and dataset-specific) gene markers</li>
<li>Compare clusters with previously identified cell types</li>
<li>Visualize clusters and gene expression using t-SNE and UMAP</li>
</ul>
</div>
<div id="usage" class="section level2">
<h2>Usage</h2>
<p>We have now made a <a href="https://welch-lab.github.io/liger/">documentation website for
rliger 2.0.0</a>. Please check it out for detailed introduction.</p>
<p>We have made a number of vignettes for typical types of analysis that
can be performed with LIGER.</p>
<ul>
<li><a href="https://welch-lab.github.io/liger/articles/Integrating_multi_scRNA_data.html">Integrating
Multiple Single-Cell RNA-seq Datasets</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/Integrating_scRNA_and_scATAC_data.html">Jointly
Defining Cell Types from scRNA-seq and scATAC-seq</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/online_iNMF_tutorial.html">Iterative
Single-Cell Multi-Omic Integration Using Online iNMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/UINMF_vignette.html">Integrating
unshared features with UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/STARmap_dropviz_vig.html">Integrating
spatial transcriptomic and transcriptomic datasets using UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/SNAREseq_walkthrough.html">scATAC
and scRNA Integration using unshared features (UINMF)</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/cross_species_vig.html">Cross-species
Analysis with UINMF</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/rna_methylation.html">Jointly
Defining Cell Types from Single-Cell RNA-seq and DNA
Methylation</a></li>
</ul>
<p>Meanwhile, since version 2.0.0, LIGER is massively updated for
usability and interoperability with other packages. Below are links to
the introduction of new features.</p>
<ul>
<li><a href="https://welch-lab.github.io/liger/articles/liger_object.html">Introduction
to new liger object and other related classes</a></li>
<li><a href="https://welch-lab.github.io/liger/articles/liger_with_seurat.html">Running
Liger directly on Seurat objects</a></li>
</ul>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
